x-pack/plugin/core/src/main/java/org/elasticsearch/snapshots/SourceOnlySnapshot.java

```diff
@@ -270,4 +295,172 @@ private boolean assertLiveDocs(Bits liveDocs, int deletes) {
+        public void deleteFile(String name) throws IOException {
+            final Directory directory = linkedFiles.remove(name);
+            if (directory == null) {
+                wrapped.deleteFile(name);
+            } else {
+                try {
+                    wrapped.deleteFile(name);
+                } catch (NoSuchFileException | FileNotFoundException e) {
+                    // ignore
+                } finally {
+                    directory.close();
+                }
+            }
+        }
+
+        @Override
+        public long fileLength(String name) throws IOException {
+            final Directory linkedDir = linkedFiles.get(name);
+            if (linkedDir != null) {
+                return linkedDir.fileLength(name);
+            } else {
+                return wrapped.fileLength(name);
+            }
+        }
+
+        @Override
+        public IndexOutput createOutput(String name, IOContext context) throws IOException {
+            if (linkedFiles.containsKey(name)) {
+                throw new IllegalArgumentException();
+            } else {
+                return wrapped.createOutput(name, context);
+            }
+        }
+
+        @Override
+        public IndexOutput createTempOutput(String prefix, String suffix, IOContext context) throws IOException {
+            return wrapped.createTempOutput(prefix, suffix, context);
+        }
+
+        @Override
+        public void sync(Collection<String> names) throws IOException {
+            final List<String> primaryNames = new ArrayList<>();
+
+            for (String name : names) {
+                if (linkedFiles.containsKey(name) == false) {
+                    primaryNames.add(name);
+                }
+            }
+
+            if (primaryNames.isEmpty() == false) {
+                wrapped.sync(primaryNames);
+            }
+        }
+
+        @Override
+        public void syncMetaData() throws IOException {
+            wrapped.syncMetaData();
+        }
+
+        @Override
+        public void rename(String source, String dest) throws IOException {
+            if (linkedFiles.containsKey(source) || linkedFiles.containsKey(dest)) {
+                throw new IllegalArgumentException();
+            } else {
+                wrapped.rename(source, dest);
+            }
+        }
+
+        @Override
+        public IndexInput openInput(String name, IOContext context) throws IOException {
+            final Directory linkedDir = linkedFiles.get(name);
+            if (linkedDir != null) {
+                return linkedDir.openInput(name, context);
+            } else {
+                return wrapped.openInput(name, context);
+            }
+        }
+
+        @Override
+        public Lock obtainLock(String name) throws IOException {
+            return wrapped.obtainLock(name);
+        }
+
+        @Override
+        public void close() throws IOException {
+            IOUtils.close(linkedFiles.values());
```

Comment:
NIT: Maybe make it safer/shorter via `IOUtils.close(() -> IOUtils.close(linkedFiles.values()), linkedFiles::clear, wrapped);`?